### DISCUSSION TITLE
Span<T> reinterpret cast

### DISCUSSION THREAD

**Discussion by PhilPJL** [dotnet/runtime#76169] [General] [answered]
I notice that Stephen Toub says that `Span<T>` supports reinterpret cast in [this article](https://learn.microsoft.com/en-us/archive/msdn-magazine/2018/january/csharp-all-about-span-exploring-a-new-net-mainstay).

> Spans provide a multitude of benefits beyond those already mentioned. For example, spans support the notion of reinterpret casts, meaning you can cast a `Span<byte>` to be a `Span<int>` (where the 0th index into the `Span<int>` maps to the first four bytes of the `Span<byte>`). That way if you read a buffer of bytes, you can pass it off to methods that operate on grouped bytes as ints safely and efficiently.

Is that true and if so can I have an example?  The compiler objects if I try something simple like:

```
Span<byte> s = stackalloc byte[20];
Span<int> t = (Span<int>)s;
```

**Comment by PhilPJL** [comment, marked as answer]
Ok found it.  `Span<int> t = MemoryMarshal.Cast<byte, int>(s);`

**Comment by jkotas** [comment]
Note that `MemoryMarshal.Cast` is not supported in all situations. From https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.memorymarshal.cast: This method is supported only on platforms that support misaligned memory access or when the memory block is aligned by other means.

For example, if you use `MemoryMarshal.Cast<byte, double>` and the buffer is not 8-byte aligned, you are going to see `DataMisalignedException` on Arm processors.

**Reply by tannergooding** [reply]
It might be worth us exposing some `Unsafe.SupportsMisalignedData` property or similar so that devs can better write "fast" code for the common case.

**Reply by PhilPJL** [reply]
Thanks for pointing that out.  Luckily I'm on x64 only.

**Comment by peterdettman** [comment]
If `MemoryMarshal.Cast` is used with target type TTo == typeof(byte), is that guaranteed to meet the clause "the memory block is aligned by other means"?

**Reply by stephentoub** [reply]
Yes
